sklearn/inspection/_partial_dependence.py

```diff
@@ -349,21 +368,44 @@ def partial_dependence(estimator, X, features, response_method='auto',
                 "'decision_function'. Got {}.".format(response_method)
             )
 
-    n_features = X.shape[1]
-    features = np.asarray(features, dtype=np.int32, order='C').ravel()
-    if any(not (0 <= f < n_features) for f in features):
-        raise ValueError('all features must be in [0, %d]'
-                         % (n_features - 1))
+    if isinstance(features, slice):
+        err_msg = ("Unsupported type for the parameter 'features'. The "
+                   "expected type is {}. Got {!r} instead.")
+        if hasattr(X, "iloc"):
+            specified_msg = ("a column name or an integer column indice or an "
+                             "array-like containing column names or integer "
+                             "column indices")
+        else:
+            specified_msg = ("an integer column indice or an array-like "
+                             "containing integer column indices")
+        raise ValueError(err_msg.format(specified_msg, features))
```

Comment:
If the type of the argument is wrong we should raise a `TypeError` instead of a `ValueError`.